packages/nightingale-track/src/nightingale-track.ts

```diff
@@ -289,45 +316,103 @@ class NightingaleTrack extends withManager(
           this.getSingleBaseWidth(),
           this.layoutObj?.getFeatureHeight() || 0,
           f.end ? f.end - f.start + 1 : 1,
-          this.getShape(f),
-        ),
+          this.getShape(f)
+        )
       )
       .attr(
         "transform",
         (f) =>
           `translate(${this.getXFromSeqPosition(f.start)},${
             this.layoutObj?.getFeatureYPos(f.feature) || 0
-          })`,
+          })`
       )
       .style("fill", (f) => this.getFeatureFillColor(f))
       .attr("stroke", (f) => this.getFeatureColor(f))
       .style("fill-opacity", ({ feature }) =>
-        feature.opacity ? feature.opacity : 0.9,
+        feature.opacity ? feature.opacity : 0.9
       )
       .style("stroke-opacity", ({ feature }) =>
-        feature.opacity ? feature.opacity : 0.9,
+        feature.opacity ? feature.opacity : 0.9
       );
 
+    const residueGroup = fragmentGroup
+      .selectAll("g.residue-group")
+      .data((d) =>
+        d.feature.ptms
+          ? [
+              d.feature.ptms?.map((ptm) =>
+                Object.assign({}, ptm, {
+                  feature: d.feature,
+                  position: ptm.position,
+                })
+              ),
+            ]
+          : []
+      )
+      .enter()
+      .append("g")
+      .attr("class", "residue-group");
+
     fragmentGroup
       .append("rect")
       .attr("class", "outer-rectangle feature")
       .attr("width", (f) =>
         Math.max(
           0,
-          this.getSingleBaseWidth() * (f.end ? f.end - f.start + 1 : 1),
-        ),
+          this.getSingleBaseWidth() * (f.end ? f.end - f.start + 1 : 1)
+        )
       )
       .attr("height", this.layoutObj?.getFeatureHeight() || 0)
       .attr(
         "transform",
         (f) =>
           `translate(${this.getXFromSeqPosition(f.start)},${
             this.layoutObj?.getFeatureYPos(f.feature) || 0
-          })`,
+          })`
       )
       .style("fill", "transparent")
       .attr("stroke", "transparent")
       .call(bindEvents, this);
+
+    residueGroup
+      .selectAll("g.residue")
+      .data((d) => d)
+      .enter()
+      .append("path")
+      .attr("class", (f) => `${this.getShape(f)} residue`)
+      .attr("d", (f) => {
+        let ptmLength = 1;
+        // For longer proteins, the PTMs have to be shown prominent in the first look
+        if (this.length && this.length > 500) {
+          ptmLength = this.getSingleBaseWidth() < 4 ? 4 : 1;
+        }
+        return this.featureShape.getFeatureShape(
+          this.getSingleBaseWidth() / 2, // Halve the width of the ptm residue to distinguish between each other if one follows next closely
+          this.layoutObj?.getFeatureHeight() || 0,
+          ptmLength,
+          this.getShape(f)
```

Comment:
```suggestion
        let ptmLength = 1;
        // For longer proteins, the PTMs have to be shown prominent in the first look
        if (this.length && this.length > 500) {
          ptmLength = this.getSingleBaseWidth() < 4 ? 4 : 1;
        }
        return this.featureShape.getFeatureShape(
          this.getSingleBaseWidth() / 2, // Halve the width of the ptm residue to distinguish between each other if one follows next closely
          this.layoutObj?.getFeatureHeight(f) || 0,
          ptmLength,
          this.getShape(f)
```
Can also extract into its own fn as it's used below in the refresh too.